src/arr/trove/statistics.arr

```diff
@@ -180,49 +190,65 @@ fun mode-any(l :: List<Number>) -> Number:
     ms.get(num-random(ms.length()))
   end
 end
-  
-fun stdev(l :: List) -> Number:
-  doc: ```returns the standard deviation of the list 
+
+fun variance(l :: List) -> Number:
+  doc: ```returns the variance of the list
        of numbers, or raises an error if the list is empty```
   reg-mean = mean(l)
   sq-diff = l.map(lam(k): num-expt((k - reg-mean), 2) end)
   sq-mean = mean(sq-diff)
-  num-sqrt(sq-mean)
+  sq-mean
 end
 
-fun stdev-sample(l :: List) -> Number:
-  doc: ```returns the standard deviation of the list 
+fun stdev(l :: List) -> Number:
+  doc: ```returns the standard deviation of the list
+       of numbers, or raises an error if the list is empty```
+  num-sqrt(variance(l))
+end
+
+fun variance-sample(l :: List) -> Number:
+  doc: ```returns the variance of the list
        of numbers, or raises an error if the list is empty```
   len = l.length()
   reg-mean = mean(l)
   sq-diff = l.map(lam(k): num-expt((k - reg-mean), 2) end)
   sq-mean = math.sum(sq-diff) / (len - 1)
-  num-sqrt(sq-mean)
+  sq-mean
+end
+
+fun stdev-sample(l :: List) -> Number:
+  doc: ```returns the standard deviation of the list
+       of numbers, or raises an error if the list is empty```
+  num-sqrt(variance-sample(l))
 end
 
-fun linear-regression(x :: List<Number>, y :: List<Number>) -> (Number -> Number):
-  doc: "returns a linear predictor function calculated with ordinary least squares regression"
-  if x.length() <> y.length():
+# please see: https://online.stat.psu.edu/stat462/
+
+fun multiple-regression(x_s_s :: List<List<Number>>, y_s :: List<Number>) -> (Any -> Number):
+  doc: "multiple-regression"
+  intercepts = MX.col-matrix.make(raw-array-of(1, y_s.length()))
+  mx_X = intercepts.augment(MX.lists-to-matrix(x_s_s))
+  mx_Y = MX.col-matrix.make(raw-array-from-list(y_s))
+  B = MX.mtx-to-vector(MX.mtx-least-squares-solve(mx_X, mx_Y))
+  fun B_pred_fn(x_s :: List<Number>) -> Number:
```

Comment:
This is an opportunity to improve the error message here.  Create the vector, and check that its size matches B's size (which you can memoize and close over), or else `raise "This regression expects # variables, but received ## instead"`.  The alternative is going to expose the error message from `dot` (https://github.com/brownplt/pyret-lang/blob/horizon/src/arr/trove/matrices.arr#L196), which involves vocabulary that isn't meaningful to the students using this...

Reply:
I expect @schanzer will wrap this functionality _anyway_ when generating the BS:* libraries.

You might also try to `raise(E.arity-mismatch(Srcloc.builtin("matrices"), B's size, x_s' size, false))` to raise a standard Pyret arity error, and see what that looks like.

Reply:
I expect @schanzer will wrap this functionality _anyway_ when generating the BS:* libraries.

You might also try to `raise(E.arity-mismatch(Srcloc.builtin("matrices"), B's size, x_s' size, false))` to raise a standard Pyret arity error, and see what that looks like.

Reply:
Done³.  Used a basic `raise(E.message-exception(...))` as I couldn't get the other one to work and I don't see existing usage

Reply:
Probably fine -- after all, this function genuinely has an arity of 1, not an arity of N... :)